### PR TITLE
Caclulate and store album duration in music db and make available for smart playlists, jsonrpc and as an infolabel

### PR DESCRIPTION
**Pull request by the-black-eagle**
## Description
Calculates and stores the duration of an album based upon the duration of the songs off an album in a users library (some users may not have all songs) and makes the information available via an info label and the json-rpc.  It can also be used in albums smart playlists in a similar way to song duration can already be used.

- Bumps music database to MyMusic80
- Adds "albumduration" to JSON  `audioLibrary.GetAlbums` and `audioLibrary.GetAlbumDetails` methods
- Adds infolabel Listitem.Property(Album_Duration) and populates ListItem.Duration when item list content is "albums"

## Motivation and Context
The run-time of an album is generally part of the information on the rear of a cd case and is scrapable from music information sites however I found that when I wanted to add that information to the musicinfo dialog for an album that it wasn't readily available. To get that information involves getting all the songs into a container and then getting the duration of that.  Even then, although the information can be displayed, it's still not possible to find all albums with a run time of 30~40 minutes or separate out those albums that play for over two hours.

## How Has This Been Tested?
Tested locally by both upgrading old databases and creating fresh ones with MySQL and SQLite.  New listItem tested by modifying the musicinfo dialog to show the run time of albums.  Created several smart playlists based around different run times.

## User related:
Adds the ability to use the duration of an album in smart playlists in the same way that the duration of songs can already be used.

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [X] All new and existing tests passed

### CODE REVIEW

**Comment by DaveTBlake** [comment]
I don't think `CMusicInfoTag.m_iAlbumDuration`  and associated methods are needed,  can reuse `m_iDuration` much like mood, ratings and compilation in song and album use the same tag fields.

Also considering LISTITEM_ALBUM_DURATION. If the listitem is an album then existing info label LISTITEM_DURATION can hold the duration (currently only song listitems have a value here). If the item is a song the data comes from songview, and vaues of the album fields of that view are stored in the item properties "album_duration" accessed by skins via  `ListItem.Property(Album_Duration)`. There may be some odd issues with when `ListItem.Property(Album_XXX)`  get populated and when they don't but that is something to fix for all the properties rather than cherry pick album duration as special.

You could share the Estuary mods to see album duration as a 3rd commit (useful for testing, and possibly included as a skin update anyway).

Note for myself over making duration a denormalised field of album table rather than a aggregate subquery in albumview and songview: 
a) such correlated subqueries are slow in MySQL because it does not use any indexes
b) possibility in future to make this scraped album data rather than depend on how many songs from the album the user has in their library.

**Comment by the-black-eagle** [comment]
After due consideration, a full independent ListItem is probably not needed (unless you'd like to show and sort by the album duration in a playlist) and a property will suffice.  Therefore I've upated the PR accordingly and added the necessary code to Estuary as a seperate commit as you suggested.

File: xbmc/GUIInfoManager.cpp [hunk 1]
```diff
@@ -6480,6 +6480,12 @@ const infomap container_str[]  = {{ "property",         CONTAINER_PROPERTY },
 ///     @return The Musicbrainz release status of the album (offical, bootleg, promotion etc)
 ///     <p><hr>
 ///     @skinning_v19 **[New Infolabel]** \link ListItem_AlbumStatus `ListItem.AlbumStatus`\endlink
+///   \table_row3{   <b>`ListItem.Property(Album_Duration)`</b>,
+///                  \anchor ListItem_Property_Album_Duration
+///                  _string_,
+///     @return The duration of the album in HH:MM:SS.
+///     <p><hr>
+///     @skinning_v19 **[New Infolabel]** \link ListItem_Property_Album_Duration `ListItem.Property(Album_Duration)`\endlink
```
Comment (DaveTBlake) [inline xbmc/GUIInfoManager.cpp:6488]: Best keep properties together, move this to just below ListItem.Property(Album_Isboxset)

File: xbmc/music/tags/MusicInfoTag.h [hunk 2]
```diff
@@ -249,6 +249,7 @@ class CMusicInfoTag final : public IArchivable, public ISerializable, public ISo
   int m_samplerate;
   int m_channels;
   int m_bitrate;
+  int m_iAlbumDuration;
```
Comment (DaveTBlake) [inline xbmc/music/tags/MusicInfoTag.h:252]: Unused property now using m_iDuration

**Review by DaveTBlake — CHANGES_REQUESTED** [review]
A quick visual pass though picked up some minors, yet to test it out. Perhaps kick off a Win64 text build for me once those are done.

File: xbmc/GUIInfoManager.cpp (hunk 1 above)
Comment (the-black-eagle) [reply xbmc/GUIInfoManager.cpp]: Moved

File: xbmc/music/tags/MusicInfoTag.h (hunk 2 above)
Comment (the-black-eagle) [reply xbmc/music/tags/MusicInfoTag.h]: Removed

**Comment by the-black-eagle** [comment]
Test build triggered :)

**Comment by zag2me** [comment]
Any performance issues with this on large libraries? Does it take long to calculate the album duration?

**Comment by the-black-eagle** [comment]
@zag2me I guess that depends on the size of your db.  The initial calculations are done as part of the db upgrade migration.  I ran the upgrade query against a db with 1327 albums and 26017 tracks.

`Result: query executed successfully. Took 596ms, 1327 rows affected` 

 So just over half a second.  When adding stuff the duration is calculated as each track is added to the db so the difference there will be negligible.

**Review by DaveTBlake — APPROVED** [review]
Finally made time to test this, sorry for delay. LGTM

JSON **audioLibrary.GetAlbums** and **audioLibrary.GetAlbumDetails**  methods tested, for example 
```
{"jsonrpc":"2.0","id":1,"method":"audioLibrary.GetAlbums", "properties":["albumduration"]}}
```

It could be made clearer that this PR actually modifies Estuary to use the new infolabel
Perhaps change the commit message to "[Estuary]Add album duration to album info dialog"

**Comment by the-black-eagle** [comment]
@DaveTBlake Commit message changed and the sql for the upgrade migration tweaked as discussed on slack.

**Comment by DaveTBlake** [comment]
Thanks, all good to go once jenkins is green again.

**Pull request merged**